src/computational_graph/graph.jl

```diff
@@ -1,338 +1,368 @@
+```
+"""
+function feynman_diagram(vertices::Vector{OperatorProduct}, topology::Vector{Vector{Int}};
+    external::Union{Nothing,AbstractVector}=nothing, factor=one(_dtype.factor), weight=zero(_dtype.weight), name="", type=:generic)
+
+    operators = [o for v in vertices for o in v.operators]
+    contraction = collect(Iterators.flatten(topology))
+    if isnothing(external)
+        external = [i for i in eachindex(operators) if i ∉ contraction]
+    end
+    @assert length(unique(contraction)) == length(contraction) # no repeated index
+    @assert length(unique(external)) == length(external) # no repeated index
+    @assert Set(union(external, contraction)) == Set(eachindex(operators)) # external + permutation must exhaust all operators
+
+    permutation = union(contraction, external)
+    _external = intersect(external, contraction)
+
+    fermionic_operators = isfermionic.(operators)
+    filter!(p -> fermionic_operators[p], permutation)
+    sign = isempty(permutation) ? 1 : parity(sortperm(permutation))
+
+    filter!(p -> fermionic_operators[p], _external)
+    ext_sign = isempty(_external) ? 1 : parity(sortperm(_external))
+    # println(_external, ", ", ext_sign)
+
+    subgraphs = [propagator(reduce(*, operators[connection])) for connection in topology]
+    g = Graph(vertices; external=external, subgraphs=subgraphs, topology=topology, name=name,
+        type=type, operator=Prod(), factor=factor * sign * ext_sign, weight=weight)
+    return g
+end
+
+"""
+    function propagator(ops::OperatorProduct;
+        name="", diagtype=:propagator, factor=one(_dtype.factor), weight=zero(_dtype.weight), operator=Sum())
+
+    Create a propagator-type Graph from given OperatorProduct `ops`.
+"""
+function propagator(ops::OperatorProduct;
+    name="", diagtype=:propagator, factor=one(_dtype.factor), weight=zero(_dtype.weight), operator=Sum())
+    return Graph([ops,]; external=collect(eachindex(ops)), type=diagtype, name=name, operator=operator, factor=factor, weight=weight)
+end
+
+"""
+    function standardize_order!(g::Graph)
+
+    Standardize the order of all leaves (propagators) of Graph by correlator ordering.
+
+# Example: 
+```julia-repl
+julia> g = propagator(𝑓⁺(1)𝑏⁺(2)𝜙(3)𝑓⁻(1)𝑏⁻(2))
+1: propagator graph from f⁺(1)b⁺(2)ϕ(3)f⁻(1)b⁻(2)
+
+julia> standardize_order!(g)
+
+julia> g, g.factor
+(1: propagator graph from f⁻(1)b⁻(2)ϕ(3)b⁺(2)f⁺(1), -1.0)
+```
+"""
+function standardize_order!(g::Graph)
+    for node in PreOrderDFS(g)
+        if isempty(node.subgraphs)
+            sign, perm = correlator_order(OperatorProduct(external(node)))
+            node.external = node.external[perm]
+            node.factor *= sign
+        else
+            sign, perm = normal_order(OperatorProduct(external(node)))
+            node.external = node.external[perm]
+            node.factor *= sign
+        end
+    end
+end
+
+prune_unary(g::Graph)=((length(g.subgraph)==1 && g.subgraph_factors[1]==1 && g.factor == 1) ? g.subgraph[1] : g)

```

Comment:
This function only makes sense for Prod and Sum operations; it encodes that (+g) = g and (*g) = g.
So, we should convert it to a long-form function and test that `g.operator in [Prod, Sum]`.